### ISSUE TITLE
unable to read -> (modify) -> write due to container_source

### ISSUE THREAD

**Issue by NileGraddis** [hdmf-dev/hdmf#315]
problem
-----------

I would like to read an nwbfile in, potentially modify the nwbfile object, then write it out to some other sink (maybe a new h5 file). This fails because container source cannot be reset. 

repro
-------
```Python
from datetime import datetime

import pynwb

def main():

    with pynwb.NWBHDF5IO(path="repro.nwb", mode="w") as nwb_writer:
        nwbfile = pynwb.NWBFile(
            session_description="",
            identifier="",
            session_start_time=datetime.now()
        )
        nwb_writer.write(nwbfile)

    with pynwb.NWBHDF5IO(path="repro.nwb", mode="r") as nwb_reader:
        nwbfile_obt = nwb_reader.read()
        
        with pynwb.NWBHDF5IO(path="repro_out.nwb", mode="w") as nwb_rewriter:
            nwb_rewriter.write(nwbfile_obt)

if __name__ == "__main__":
    main()
```
produces:
```
Traceback (most recent call last):
  File "container_source_repro.py", line 25, in <module>
    main()
  File "container_source_repro.py", line 20, in main
    nwb_rewriter.write(nwbfile_obt)
  File "/home/nile/Documents/work/allen_institute/hdmf/src/hdmf/utils.py", line 483, in func_call
    return func(args[0], **parsed['args'])
  File "/home/nile/Documents/work/allen_institute/hdmf/src/hdmf/backends/hdf5/h5tools.py", line 280, in write
    call_docval_func(super().write, kwargs)
  File "/home/nile/Documents/work/allen_institute/hdmf/src/hdmf/utils.py", line 364, in call_docval_func
    return func(*fargs, **fkwargs)
  File "/home/nile/Documents/work/allen_institute/hdmf/src/hdmf/utils.py", line 483, in func_call
    return func(args[0], **parsed['args'])
  File "/home/nile/Documents/work/allen_institute/hdmf/src/hdmf/backends/io.py", line 43, in write
    f_builder = self.__manager.build(container, source=self.__source)
  File "/home/nile/Documents/work/allen_institute/hdmf/src/hdmf/utils.py", line 483, in func_call
    return func(args[0], **parsed['args'])
  File "/home/nile/Documents/work/allen_institute/hdmf/src/hdmf/build/manager.py", line 157, in build
    container.__class__.__name__))
ValueError: Cannot change container_source once set: 'root' pynwb.file.NWBFile
```

with pynwb 1.3.0, hdmf 1.6.1 in python 3.7.6 on ubuntu 19.10.

more
-------
It seems like this workflow would be handy for appending data/metadata or converting to new backends. Obviously, several workarounds exist: we can copy the file and open it `"r+"`, we can traverse the nwbfile and set all of the container sources directly via private attribute. These are not great options, since they are either fragile or inflexible.

**Comment by ajtritt** [comment]
@oruebel @rly and I chatted about this. We kicked around a couple solutions and think the cleanest is for HDMF to provide an export functionality, which writes a container to the new sink and then append to that sink. 

This would need a new `export` function that looks something like this:

```python
def export(container, hdmfio, **write_kwargs):
    """
    Args:
        container:          the container to copy to the new sink
        hdmfio:             the sink to copy the container to
        write_kwargs:       kwargs to be passed into hdmfio.write
    """
    ...
``` 

Your snippet would look like this instead:

```Python
from datetime import datetime

import pynwb

def main():

    with pynwb.NWBHDF5IO(path="repro.nwb", mode="w") as nwb_writer:
        nwbfile = pynwb.NWBFile(
            session_description="",
            identifier="",
            session_start_time=datetime.now()
        )
        nwb_writer.write(nwbfile)

    with pynwb.NWBHDF5IO(path="repro.nwb", mode="r") as nwb_reader:
        nwbfile_obt = nwb_reader.read()
        
        with pynwb.NWBHDF5IO(path="repro_out.nwb", mode="w") as nwb_rewriter:
            pynwb.export(nwbfile_obt, nwb_rewriter)

if __name__ == "__main__":
    main()
```

**Comment by rly** [comment]
This has been fixed by #388.
